File: app/src/main/java/com/xnote/app/domain/model/AppSettings.kt
package com.xnote.app.domain.model

// -- Type Definitions

enum class ThemeMode {
    System,
    Light,
    Dark,
}

enum class AppFontSize(val scale: Float) { Standard(1f), Large(1.15f), ExtraLarge(1.3f) }

enum class ReadingLayout(val widthDp: Int, val lineHeightScale: Float) {
    Compact(560, 1f), Standard(840, 1f), Relaxed(640, 1.25f),
}

data class AppSettings(
    val defaultBackground: BackgroundKey,
    val themeMode: ThemeMode,
    val markdownShortcutsEnabled: Boolean,
    val reduceMotion: Boolean = false,
    val highContrast: Boolean = false,
    val fontSize: AppFontSize = AppFontSize.Standard,
    val readingLayout: ReadingLayout = ReadingLayout.Standard,
)

// -- Functions

fun defaultAppSettings(): AppSettings = AppSettings(
    defaultBackground = defaultBackgroundKey(),
    themeMode = ThemeMode.System,
    markdownShortcutsEnabled = true,
)

File: flutter/lib/domain/model/app_settings.dart
import 'background_key.dart';

// -- Type Definitions

enum AppThemeMode { system, light, dark }

final class AppSettings {
  const AppSettings({
    required this.defaultBackground,
    required this.themeMode,
  });

  final BackgroundKey defaultBackground;
  final AppThemeMode themeMode;

  // -- Functions

  AppSettings copyWith({
    BackgroundKey? defaultBackground,
    AppThemeMode? themeMode,
  }) {
    return AppSettings(
      defaultBackground: defaultBackground ?? this.defaultBackground,
      themeMode: themeMode ?? this.themeMode,
    );
  }

  @override
  bool operator ==(Object other) {
    return other is AppSettings &&
        other.defaultBackground == defaultBackground &&
        other.themeMode == themeMode;
  }

  @override
  int get hashCode => Object.hash(defaultBackground, themeMode);
}

// -- Functions

AppSettings defaultAppSettings() => AppSettings(
      defaultBackground: defaultBackgroundKey(),
      themeMode: AppThemeMode.system,
    );
